src/core/projectinfo.cpp

```diff
@@ -324,6 +324,42 @@ void ProjectInfo::saveLayerSnappingConfiguration( QgsMapLayer *layer )
   mSettings.endGroup();
 }
 
+void ProjectInfo::saveLayerRememberedFields( QgsMapLayer *layer )
+{
+  if ( mFilePath.isEmpty() )
+    return;
+
+  QgsVectorLayer *vlayer = qobject_cast<QgsVectorLayer *>( layer );
+  if ( !vlayer )
+    return;
+
+  QVariantMap rememberedFields;
+  QgsEditFormConfig config = vlayer->editFormConfig();
+  const QgsFields fields = vlayer->fields();
+  for ( int i = 0; i < fields.size(); i++ )
+  {
+    rememberedFields.insert( fields.at( i ).name(), config.reuseLastValue( i ) );
+  }
+
+  const bool isDataset = QgsProject::instance()->readBoolEntry( QStringLiteral( "QField" ), QStringLiteral( "isDataset" ), false );
+
+  // Prefix id with :: to avoid loss of slash on linux paths
+  QString id( QStringLiteral( "::" ) );
+  if ( isDataset )
+  {
+    // For non-project datasets, the layer id is random, use the source URI
+    id += layer->source();
```

Comment:
I thought when importing a layer as dataset it always had a different path. Good to learn it's not that bad!